### PR TITLE
[REVIEW] cyber foundation notebook and scripts 

### PR DESCRIPTION
**Pull request by gbatmaz**
## Description
Cyber foundation notebook, model and scripts that can learn from raw cyber logs and generate synthetic logs which can be used for various downstream tasks

### CODE REVIEW

File: cyber-foundation/dataset/azure-ad-logs-sample-training-data.json
Comment (hsin-c) [inline cyber-foundation/dataset/azure-ad-logs-sample-training-data.json:1]: Add a readme for the dataset in this same folder (can copy from morpheus proper)

File: cyber-foundation/training-inference/train.py [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+    if it > lr_decay_iters:
+        return min_lr
+    # 3) in between, use cosine decay down to min learning rate
+    decay_ratio = (it - warmup_iters) / (lr_decay_iters - warmup_iters)
+    assert 0 <= decay_ratio <= 1
+    coeff = 0.5 * (1.0 + math.cos(math.pi * decay_ratio))  # coeff ranges 0..1
+    return min_lr + coeff * (learning_rate - min_lr)
+
+
+# # logging
```
Comment (tzemicheal) [inline cyber-foundation/training-inference/train.py:264]: Clean unused comments.

File: cyber-foundation/training-inference/prepare.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+print("all the unique characters:", ''.join(chars))
+print(f"vocab size: {vocab_size:,}")
+
+# create a mapping from characters to integers
+stoi = {ch: i for i, ch in enumerate(chars)}
+itos = {i: ch for i, ch in enumerate(chars)}
+
+
+def encode(s):
+    return [stoi[c] for c in s]  # encoder: take a string, output a list of integers
```
Comment (tzemicheal) [inline cyber-foundation/training-inference/prepare.py:68]: -rearrange the comment "encoder: takes ..." above the `return` line

File: cyber-foundation/training-inference/prepare.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+
+def decode(sequence):
+    return ''.join([itos[i] for i in sequence])  # decoder: take a list of integers, output a string
```
Comment (tzemicheal) [inline cyber-foundation/training-inference/prepare.py:72]: Same like above

**Review by tzemicheal — CHANGES_REQUESTED** [review]
Overall, the code LGTM! I understand the usecase depends on third party codes. Would be nice if the comments are rearranged for some of the codes in the scripts.

File: cyber-foundation/dataset/azure-ad-logs-sample-training-data.json
Comment (gbatmaz) [reply cyber-foundation/dataset/azure-ad-logs-sample-training-data.json:1]: Done Thank you

File: cyber-foundation/training-inference/train.py (hunk 1 above)
Comment (tzemicheal) [reply cyber-foundation/training-inference/train.py:264]: Will be updated later

**Review by tzemicheal — APPROVED** [review]

File: cyber-foundation/README.md [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+
+* This model is intended for developers who want to build GPT based synthetic log generator
+
+### Name who is intended to benefit from this model. 
+
+* The intended beneficiaries of this model are developers who aim to generate synthetic Azure logs.
+
+### Describe the model output. 
+
+* This model output can be used as a binary result, DGA or Benign
```
Comment (hsin-c) [inline cyber-foundation/README.md:200]: Is this off?

File: cyber-foundation/dataset/README.md [hunk 5]
```diff
@@ -0,0 +1,14 @@
+## Cyber Foundation Model Data
```
Comment (hsin-c) [inline cyber-foundation/dataset/README.md:1]: We could add a link to morpheus proper too for this identical dataset

File: cyber-foundation/training-inference/cyber-foundation.ipynb [hunk 6]
```diff
@@ -0,0 +1,1318 @@
...
+    "\n",
+    "By leveraging these models, users can enhance their cybersecurity efforts and improve overall system resilience.\n",
+    "\n",
+    "The associated scripts are taken from this [repo](https://github.com/karpathy/nanoGPT). We chose this repo because it has very few dependencies and allows us to easily change the model size and hyperparameters.\n",
+    "\n",
+    "This notebook shows that a small GPT2 size model is enough to learn the specifics of raw cyber logs and generate realistic synthetic logs. Users could use this approach for potential downstream tasks:\n",
+    "1. Generating realistic training data: \n",
+    "   The structure of logs can differ from setting to setting. A model trained with propriety data can generate realistic logs true to the setting and incorporate them into future model retraining cycles, helping reduce false positives. \n",
+    "\n",
+    "2. Generating attack behaviour to test detectors:**\n",
```
Comment (hsin-c) [inline cyber-foundation/training-inference/cyber-foundation.ipynb:58]: Is this extra **?

File: cyber-foundation/training-inference/cyber-foundation.ipynb [hunk 7]
```diff
@@ -0,0 +1,1318 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "ecc88e04-7f2f-403b-8dcb-d4ba08eeb7c6",
+   "metadata": {},
+   "source": [
+    "Once prompted with the beginning of the code, the pre-trained model completes the rest of the logs. \n",
+    "\n",
+    "The sampling script has a small maximum token limit to make the notebook run faster for demonstration purposes. After each iteration, generation stops when the limit is reached, causing an incomplete log in each iteration. Such incomplete or non-intact logs should be removed with post-processing before using them for downstream tasks.\n",
+    "\n",
+    "In our experiments with real logs, 90% of the generated logs were intact.\n",
```
Comment (hsin-c) [inline cyber-foundation/training-inference/cyber-foundation.ipynb:1256]: "the generated logs were intact."-- intact sounds 
a bit misleading, like it's reproducing what's in the training set. Maybe we can say it's in parsable json format?

Comment (gbatmaz) [reply cyber-foundation/training-inference/cyber-foundation.ipynb]: Great point. Thanks for the feedback! Updated now.

**Comment by gbatmaz** [comment]
thank you both! I've addressed all the comments @hsin-c Could you please approve and merge?

**Review by hsin-c — APPROVED** [review]
LGTM

**Pull request merged**